### ISSUE TITLE
[P0] Backend base API

### ISSUE THREAD

**Issue by seiwonpark** [kookmin-sw/capstone-2023-40#31]
**(Edited)**

- user
- survey
- survey_question
- survey_auth_provider
- participation
- answered_question
- survey_answer

What should work:
**auth**
- POST `/api/auth/register` : register user
- POST `/api/auth/login` : sign in (handles with session authentication)

**users**
- GET `/api/users/profile` : fetch user profile (find user from session)
- PUT `/api/users/profile` : update user's profile (find user from session)
- DELETE `/api/users` : delete user (find user from session)

**surveys**
- GET `/api/surveys` : fetchAll survey list (Which will be shown on SurveyListPage)
- GET `/api/surveys/:id` : fetch survey with `:id` and its related questions
- POST `/api/surveys` : create survey with `questions`s and `question_bank`s
- POST `/api/surveys/submit` : submit survey answers
- PUT `/api/surveys` : modify survey

**Comment by bezzang-dev** [comment]
What do you think about renaming the name of '**survey_item**' table to '**question**'??
I think this name is more intuitive, and if we make a separate table for the choice(attribute in survey_item table), there might be some confusion with name. (ex.survey_item and survey_choice..?)
and then I think we should rename **answered_survey_item** to **answered_question** as well.

**Comment by seiwonpark** [comment]
Sure `question` looks more neat. Intuitive, simple, perfect!

**Comment by seiwonpark** [comment]
And I've checked a little bit of your code. Could you add `Column()`s with `snake_case_column_name` on every column? We could add additional options and to keep consistency

Additionally could you modify every `id`s with its entity name?

ex.
```java
Column(name = "user_id")   // since `id` could be confusing
private Long userId;
```

**Comment by bezzang-dev** [comment]
Okay, I will!

**Comment by bezzang-dev** [comment]
- GET `/survey/:id `: fetchAll surveys with :id and its related questions

Could you explain more about what this means??
I didn't quite understand fetch all the survey with id because id is unique... 

its means getting for a survey with id and its related questions?

**Comment by seiwonpark** [comment]
Oh I'm sorry for the misleading words. 

Yeah it should return all related questions in that **specific** survey with `{id}` and the **specific** survey itself as well.

I've updated it! Thank you!

**Comment by bezzang-dev** [comment]
Thank you for your explanation!

**Comment by bezzang-dev** [comment]
`POST /survey : create survey with questions`

While implementing , I thought that if Client created a survey, Server should save the survey with the user information.

Then, when the `POST` request comes in, how will the user data come in response body? Can data about two entities (User, Survey)be received in one Http request? 
If We use the session authentication, does the session enter the response body and get user information from it?

**Comment by seiwonpark** [comment]
Think about some blogs like tistory, or naver blog. We don't actually put email or something else in a form but just send the contents.

As you explained, when `POST` request is sent, then the server should find that user from session. But we don't have to care about the **response body**. On every single http request, server should verify the user(querying happens here) and validate the request, and finally handle the request.(querying happens here)

(`JSESSIONID` is sent inside Cookie. And the cookie is in the Header of http request)

I don't really get what you meant by `response body`. We take say `title`, `content` for survey and `JSESSIONID` for user authentication, and that's it.

**Comment by bezzang-dev** [comment]
Thank you for the explanation.

The reason I asked you this question is because
When the client sends a request to create a survey,
I don't know how to save `user_id` in the participation table.
Because when a survey is created, survey_id is automatically created, but the user_id is not a new value, but a value that already exists.

I need to know the user_id who wrote this survey, but I don't know how to find it…

**Comment by seiwonpark** [comment]
So we have to query to find the user using session id which is sent through cookie in http request header.

This is why we need to balance the request since on every request server should find out `Who sent this request?`
And after the server find the user, then we can take it.

Here's an example code for finding `Who the hack sent this request?`

```java
// 1. find JSESSIONID from Header
 String sessionId = request.getHeader("JSESSIONID");
 
 // 2. See if the session is authenticated and valid
 Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
 
 // 3. then we can get user_id or username, etc...
```

And this should be run on every single HTTP Request. Imagine how much loads should be handled in our poor single server.

Or if you don't like this, do you want to go back token authentication?

**Comment by bezzang-dev** [comment]
I didn't think of a way to get user information through session ID. Now I understand. Thank you for the explanation.
(and I want to proceed with session authentication...!)

**Comment by seiwonpark** [comment]
Closes due to duplicated issue at #64. And User API has been implemented at #60 and #61.
